### DISCUSSION TITLE
Can before_all be used with Rails' parallelize?

### DISCUSSION THREAD

**Discussion by peret** [test-prof/test-prof#211] [Q&A]
I'm refactoring a large Rails test suite (Minitest) and had some success with using the `before_all` hook to cut down on expensive object fabrication 👍 
However, when running the test suite now using Rails' build-in parallelization feature, `parallelize`, I get the message `!!! before_all transaction has been already rollbacked and could work incorrectly` at the beginning. And indeed it seems that some test cases are failing due to that error.
I'm running 8 parallel workers and I get this message 7 times, so I'm assuming that all 8 processes are using the same transaction and only one of them is able to roll it back successfully. Unfortunately, I don't know enough about how Rails `parallelize` works in conjunction with the `before_all` hook. E.g. in my understanding, Rails creates a separate DB (+ connection?) for each worker, so how could they  interfere with each other?

Does anyone have experience with this and can tell me how to get it runnning or if it's even possible?

**Comment by palkan** [comment]
>  if it's even possible?

I think, it is.
Since parallel executors uses different databases, we need to open a transaction per database.

It seems feasible: we can keep track of the current worker number in `#activate`: https://github.com/test-prof/test-prof/blob/9beed74d278e371eb6589958b5498de87b41f342/lib/test_prof/recipes/minitest/before_all.rb#L27-L36

We can use `Process.pid` to identify the current executor.

**Reply by peret** [reply]
After your reply I digged into it more and checked the Rails and minitest code. The problem is actually that the parallel executors of minitest and Rails override the `run` class method and instead of executing the test case right away, they just add them to a queue that's shared between threads/processes. That means the run method already returns when all jobs have been enqueued, before any of the tests are actually executed  and `deactivate!` is called too early:

https://github.com/test-prof/test-prof/blob/9beed74d278e371eb6589958b5498de87b41f342/lib/test_prof/recipes/minitest/before_all.rb#L88-L95

It seems I have found a solution to work around that and will prepare a PR for further discussion :)
Also, I should maybe clarify, I'm using process-parallelization and the transactions are actually separate already, so no problem there, at least when using processes.
